cEP-2.md

```diff
@@ -0,0 +1,33 @@
+#Next Gen Bear Design
+| Metadata     | 					                        |
+| ------------ |-----------------------------------------|
+| cEP          | 2                                       |
+| Version      | 2.1                                     |  
+| Title        | Next gen bear design                    |
+| Authors	     | Udayan Tandon <udayan12167@iiitd.ac.in> |
+| Status       | Active                                  |
+| Type         | Process                                 |
+#Abstract
+This document describes how the next generation bears will be designed and implemented.
+
+This cEP describes a new implementation of bears in coala which should make them simpler and make their design more intuitive and better.
+
+#Introduction
+Currently the bear design is clunky and bad. The current design makes global bears dependant on local bears through it's queuing mechanism. It uses communication signals over a queue to indicate that global bears are done with their execution and local bears can begin. Also we have a permanent logger thread that we can do away with.
+
+The new proposal aims to do away with the queuing control signals. Utilising a multiprocessing pool to generate the results from the bears. No longer will the local bears be dependent on global bears to finish.
+
+#Proposed Change
+Here is the detailed implementation stepwise:
+
+1. coala will have a core multiprocessing pool.
+2. A bear will get two new methods. `execute_tasks()` and `generate_tasks()`.
+	* The `generate_tasks()` is resposible for creating method calls with the correct arguments.
+	* The `execute_tasks()` method is responsible for giving the generated method calls to the multiprocessing pool.
+3. Each bear will load the core coala pool with different types of tasks.
+4. Global bears fire tasks that process all files.
+5. Local bears fire tasks for each individual file.
+6. Once the tasks are completed the results are put in a queue.
+7. This queue is polled by the main thread constantly to log the results and handle the interactions.
+
+A prototype of the above proposal can be found: <https://github.com/coala-analyzer/next-gen-bears-prototype> 
```

Comment:
make meaningful examples out of that, this repository is just scratched code for experimental and concept-design purposes, don't link to that^^